src/codec.rs

```diff
@@ -253,8 +258,16 @@ pub trait Decode: Sized {
 	#[doc(hidden)]
 	const IS_U8: IsU8 = IsU8::No;
 
+	/// Attempt to deserialise the value from input while having already preallocated space.
+	///
+	/// Use this when you want to decode while having already preallocated space for later data in
+	/// the input. This allow preallocated space to be bounded by input len or a default maximum.
+	fn decode_inner<I: Input>(input: &mut I, preallocated: usize) -> Result<Self, Error>;
```

Comment:
I don't understand this? What is preallocated size? Given your implementation this is just an upper bound.

I thought we would solve this by using the length of the input?

Reply:
preallocated is the allocation the caller made for future data in input.

in a Vec I call decode of an element with `preallocated = (vec.capacity() - vec.len() - 1) * size_of::<T>()`

that means that inside the decoding of this element it will be aware the calle has already preallocated some space, thus if we want to guarantee `preallocated <= input_len` the decode knows that he only have `input_len - preallocated` available to pre allocate.

in decode of Vec<T> it is legal to pre-allocated up to:
```rust
let max_for_current_vec_preallocation = input.min_len().max(DEFAULT_MAX_DECODE_PREALLOCATION_SIZE)
	.saturating_sub(preallocated_by_caller);
```

Reply:
note that this could have been done the other way around:

> caller send to decode what is the maximum pre-allocation callee is allowed, callee will have to respect this number.
This would be done with having decode(input, Some(maximum)) allowing user to specify a maxium to preallocate or no maximum if wanted. Actually this might be better, maybe trusted input would be without limit on preallocation.

But all in all that doesn't change that much, it will just be like:
`decode_inner(input, pre_allocated_limit: usize)`